pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -1071,7 +1071,12 @@ func applyUpstreamTLSSettings(opts *buildClusterOpts, tls *networking.ClientTLSS
 		if c.Http2ProtocolOptions != nil {
 			// This is HTTP/2 in-mesh cluster, advertise it with ALPN.
 			if tls.Mode == networking.ClientTLSSettings_ISTIO_MUTUAL {
-				tlsContext.CommonTlsContext.AlpnProtocols = util.ALPNInMeshH2
+				// This is in-mesh cluster, advertise it with ALPN.
+				if util.IsTCPMetadataExchangeEnabled(node) {
```

Comment:
AFAIK, IsTCPMetadataExchangeEnabled() only checks the current node, how do we know the peer node is enabled for TCP metadata exchange as well?

Reply:
We don't know before sending the request out.

So, the contract is that if both current node and peer are enabled for tcp metadata exchange, they will agree on `istio-peer-exchange` alpn protocol and then tcp metadata exchange filters will exchange peer metadata.

Reply:
what if the peer node is not enabled for tcp metadata exchange? Is `istio-peer-exchange` just ignored in the peer side?

Reply:
Yeah, then whatever is the next alpn, they will agree on that alpn and tcp metadata exchange filter won't do anything on this node.

Reply:
https://tinyurl.com/istio-tcp-metadataexchange -> the design doc for this

Reply:
I see, thanks.